src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Web/Debugger/WebServer.cs

```diff
@@ -0,0 +1,1022 @@
+        }
+
+        private void SetErrorInfo(int hr, string description)
+        {
+            _uiShell.Value.SetErrorInfo(hr, description, 0, null, string.Empty);
+        }
+
+        private bool VSIsRunningElevated()
+        {
+            IVsShell3 vsShell = (IVsShell3)_vsShell.Value;
+
+            vsShell.IsRunningElevated(out bool bIsElevated);
+            return bIsElevated;
+        }
+        private static bool DirectoriesMatch(string dir1, string dir2)
+        {
+            return string.Compare(EnsureTrailingChar(dir1, '\\'), EnsureTrailingChar(dir2, '\\'), StringComparison.OrdinalIgnoreCase) == 0;
+        }
+
+        public static bool DirectoryIsEmpty(string folderPath)
+        {
+            if (Directory.Exists(folderPath))
+            {
+                if (Directory.EnumerateFiles(folderPath).Any())
+                {
+                    return false;
+                }
+
+                foreach (string dir in Directory.EnumerateDirectories(folderPath, "*.*", SearchOption.TopDirectoryOnly))
+                {
+                    if (dir != "." && dir != "..")
+                    {
+                        return false;
+                    }
+                }
+            }
+
+            return true;
+        }
+
+        private static string EnsureTrailingChar(string s, char ch)
+        {
+            return s.Length == 0 || s[s.Length - 1] != ch ? s + ch : s;
+        }
+
+        private static void SplitUrl(string url, out string server, out string relUrl)
+        {
+            Uri uri = new Uri(url);
+            relUrl = uri.GetComponents(UriComponents.Path, UriFormat.SafeUnescaped);
+            if (!relUrl.StartsWith("/"))
+            {
+                relUrl = relUrl.Insert(0, "/");
+            }
+
+            server = uri.GetComponents(UriComponents.SchemeAndServer, UriFormat.SafeUnescaped);
+        }
+
+        private static string GetDeveloperEnvironement()
+        {
+            return GetFullEnvironmentAsString(new KeyValuePair<string, string>[] {new KeyValuePair<string, string>("DEV_ENVIRONMENT", "1")});
+        }
+
+        private static string GetFullEnvironmentAsString(IEnumerable<KeyValuePair<string, string>> envValuesToAdd)
+        {
+            // First add in the system env variables if requests
+            StringBuilder newEnvironmentValuesString = new StringBuilder();
+
+            // The IDictionary returned by the framework is case sensitive so we can't just add to the one
+            // they return. Instead, we set all the environment varaibles (and track any existing values), get the updated environment
+            // and then put the original values back. We store null or entries not found.
+            Dictionary<string, string> envValuesToReset = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase);
+            foreach (KeyValuePair<string, string> envValueToAdd in envValuesToAdd)
```

Comment:
Can use deconstruction in these foreach loops. For example:

```suggestion
            foreach ((string key, string value) in envValuesToAdd)
```

Reply:
I can for the two of the loops, but not for Environment.GertEnvironmentVariables() one. There is no deconstruction support for it.

Reply:
We actually added a `Deconstruct` extension method for `KeyValuePair<,>` (as one is not provided in .NET Framework), so you could add one for `DictionaryEntry` if you like the pattern. Given this'll move to WebTools if it gets the go-ahead, you'd need to port that extension method too.